modules/lvm2/udiskslinuxlogicalvolume.c

```diff
@@ -508,6 +515,120 @@ handle_delete (UDisksLogicalVolume   *_volume,
 
 /* ---------------------------------------------------------------------------------------------------- */
 
+static gboolean
+handle_repair (UDisksLogicalVolume   *_volume,
+               GDBusMethodInvocation *invocation,
+               const gchar *const    *arg_pvs,
+               GVariant              *options)
+{
+  GError *error = NULL;
+  UDisksLinuxLogicalVolume *volume = UDISKS_LINUX_LOGICAL_VOLUME (_volume);
+  UDisksLinuxLogicalVolumeObject *object = NULL;
+  UDisksDaemon *daemon = NULL;
+  uid_t caller_uid;
+  UDisksLinuxVolumeGroupObject *group_object;
+  LVJobData data;
+
+  if (!common_setup (volume, invocation, options,
+                     N_("Authentication is required to delete a logical volume"),
+                     &object, &daemon, &caller_uid))
+    goto out;
+
+  group_object = udisks_linux_logical_volume_object_get_volume_group (object);
+  data.vg_name = udisks_linux_volume_group_object_get_name (group_object);
+  data.lv_name = udisks_linux_logical_volume_object_get_name (object);
+
+  if (arg_pvs)
+    {
+      int n;
+
+      for (n = 0; arg_pvs[n] != NULL; n++)
+        ;
+      data.new_lv_pvs = (const gchar **)g_new(gchar *, n+1); // XXX - free this
+      for (n = 0; arg_pvs[n] != NULL; n++)
+        {
+          UDisksObject *pvol_object = NULL;
+          UDisksBlock *block = NULL;
+          UDisksPhysicalVolume *pvol = NULL;
+
+          pvol_object = udisks_daemon_find_object (daemon, arg_pvs[n]);
+          if (pvol_object == NULL)
+            {
+              g_dbus_method_invocation_return_error (invocation,
+                                                     UDISKS_ERROR,
+                                                     UDISKS_ERROR_FAILED,
+                                                     "Invalid object path %s at index %u",
+                                                     arg_pvs[n], n);
+              goto out;
+            }
+
+          block = udisks_object_get_block (pvol_object);
+          pvol = udisks_object_get_physical_volume (pvol_object);
+          if (block == NULL || pvol == NULL)
+            {
+              g_dbus_method_invocation_return_error (invocation,
+                                                     UDISKS_ERROR,
+                                                     UDISKS_ERROR_FAILED,
+                                                     "Object path %s for index %u is not a physical volume",
+                                                     arg_pvs[n], n);
+              g_object_unref (pvol_object);
```

Comment:
Leaking `block` or `pvol`. Alternatively you may use `udisks_object_peek_block()` for a weak ref.